northd/helpers.dl

```diff
@@ -28,3 +28,47 @@ SwitchRouterPeer(lsp, lsp_name, lrp) :-
     nb.Logical_Switch_Port(._uuid = lsp, .name = lsp_name, .__type = "router", .options = options),
     Some{var router_port} = map_get(options, "router-port"),
     nb.Logical_Router_Port(.name = router_port, ._uuid = lrp).
+
+function map_get_bool_def(m: Map<string, string>,
+                          k: string, def: bool): bool = {
+    match (map_get(m, k)) {
+        None -> def,
+        Some{x} -> {
+            if (def) {
+                str_to_lower(x) != "false"
+            } else {
+                str_to_lower(x) == "true"
+            }
+        }
+    }
+}
+
+function map_get_int_def(m: Map<string, string>, k: string,
+                         def: integer): integer = {
+    match (map_get(m, k)) {
+        None -> def,
+        Some{x} -> {
+            match (str_to_int(x, 10)) {
+                Some{v} -> v,
+                None -> def
+            }
+        }
+    }
+}
+
+function map_get_int_def_limit(m: Map<string, string>, k: string, def: integer,
+                               min: integer, max: integer): integer = {
+    var v = map_get_int_def(m, k, def);
+    var v1 = {
+        if (v < min) min else v
+    };
+    if (v1 > max) max else v1
+}
+
+function map_get_str_def(m: Map<string, string>, k: string,
```

Comment:
> This function can be made generic, so you don't need to re-implement it for other types:
> 
> ```
> function map_get_def(m: Map<'K, 'V>, k: 'K,
>                          def: 'V): 'V = {
>     match (map_get(m, k)) {
>         None -> def,
>         Some{x} -> x
>     }
> }
> ```
> 
> This might be an overkill, though, as I am guessing OVN only has string-to-string maps at the moment.

Indeed, the problem is that OVN maps are string-to-string, storing string representations of int, bool, string. I don't think the generic `map_get_def` would bring too much benefit because we'd need to have different parsers for different types of representations (e.g., int vs bool). Maybe there's a better way to do it. Any suggestions?

Reply:
Not really. I don't know if it's possible for OVN to rely less on strings and used more strongly typed representation.  DDlog does not have lambdas at the moment, so you cannot supply a parser function as an argument to `map_get_def`, so I guess your implementation is the best we can do right now.